contracts/gov/proposals/TokenSaleProposal.sol

```diff
@@ -0,0 +1,349 @@
+
+            saleToken.safeTransfer(
+                msg.sender,
+                vestingWithdrawAmounts[i].from18(saleToken.decimals())
+            );
+        }
+    }
+
+    function buy(uint256 tierId, address tokenToBuyWith, uint256 amount) external payable {
+        bool isNativeCurrency = tokenToBuyWith == ETHEREUM_ADDRESS;
+
+        uint256 saleTokenAmount = getSaleTokenAmount(
+            tierId,
+            tokenToBuyWith,
+            isNativeCurrency ? msg.value : amount
+        );
+
+        Tier storage tier = _tiers[tierId];
+
+        TierView memory tierView = tier.tierView;
+        TierInfo storage tierInfo = tier.tierInfo;
+
+        _amountToSell[tierView.saleTokenAddress] -= saleTokenAmount;
+
+        tierView.saleTokenAddress.sendFunds(
+            msg.sender,
+            saleTokenAmount.percentage(PERCENTAGE_100 - tierView.vestingPercentage)
+        );
+
+        tierInfo.totalSold += saleTokenAmount;
+        tierInfo.customers[msg.sender] = Purchase({
+            purchaseTime: block.timestamp,
+            vestingAmount: saleTokenAmount.percentage(tierView.vestingPercentage),
+            latestVestingWithdraw: block.timestamp
+        });
+
+        if (isNativeCurrency) {
+            (bool success, ) = govAddress.call{value: msg.value}("");
+            require(success, "TSP: failed to transfer ether");
+        } else {
+            ERC20(tokenToBuyWith).safeTransferFrom(
+                msg.sender,
+                govAddress,
+                amount.from18(ERC20(tokenToBuyWith).decimals())
+            );
+        }
+    }
+
+    function recover(uint256[] calldata tierIds) external {
+        uint256[] memory recoveringAmounts = getRecoveringAmounts(tierIds);
+
+        for (uint256 i = 0; i < recoveringAmounts.length; i++) {
+            ERC20 saleToken = ERC20(_tiers[tierIds[i]].tierView.saleTokenAddress);
+
+            saleToken.safeTransfer(govAddress, recoveringAmounts[i].from18(saleToken.decimals()));
+        }
+    }
+
+    function getSaleTokenAmount(
+        uint256 tierId,
+        address tokenToBuyWith,
+        uint256 amount
+    ) public view ifTierExists(tierId) returns (uint256) {
+        require(amount > 0, "TSP: zero amount");
+
+        require(balanceOf(msg.sender, tierId) == 1, "TSP: not whitelisted");
+
+        Tier storage tier = _tiers[tierId];
+
+        TierView memory tierView = tier.tierView;
+        TierInfo storage tierInfo = tier.tierInfo;
+
+        require(
+            tierView.saleStartTime <= block.timestamp && block.timestamp <= tierView.saleEndTime,
+            "TSP: cannot buy now"
+        );
+        require(tierInfo.customers[msg.sender].purchaseTime == 0, "TSP: cannot buy twice");
+
+        uint256 exchangeRate = tierInfo.rates[tokenToBuyWith];
+        require(exchangeRate != 0, "TSP: incorrect token");
+
+        uint256 saleTokenAmount = amount.ratio(exchangeRate, PRECISION);
+
+        require(
```

Comment:
Should include what the user has already bought

Reply:
discussed